Add support for hidden receivers in saltpack encrypt

jacobhaven opened the pull request.

maxtaco commented on go/saltpack/encrypt.go:
```diff
@@ -149,17 +150,19 @@ func (es *encryptStream) init(sender BoxSecretKey, receivers []BoxPublicKey) err
-		}
+			// Don't specify the receivers if this public key wants to hide
+			if !receiver.HideIdentity() {
+				keys.ReceiverKID = receiver.ToKID()
```
@jacobhaven note there already is support in the `go/saltpack` library for anonymous receivers via this mechanism.  The current API allows some receivers to be anonymous and some to be explicitly named.  I'm not saying we need to expose that all the way through to the CLI, but it might be at least useful to use this mechanism that's already here.

jacobhaven replied:
This would require setting creating new `BoxPublicKey`s for each receiver with a `HideIdentity() -> true` method. It's really an encryption option, not tied to the specific receivers.

maxtaco replied:
Exactly. Which wouldn't be too bad since the `naclBoxPublicKey` type is a throw-away shim around the key. So you can introduce a "supertype" in `libkb/saltpack.go` that is just of the form `hiddenNaclBoxPublicKey` and create those rather than `naclBoxPublicKey` in `SaltpackEncrypt` line 26.

jacobhaven commented:
@maxtaco I've amended this PR with your suggestion. One problem: The ability to write a unit test is to unit test this as the `engine` package doesn't have access to the `saltpack` codec decoding and header parsing. I've added an `engine.TestSaltpackAnonymousEncrypt` test, but I haven't been able to get it to connect to a local keybase server:
`common_test.go:87: API network error: Post http://localhost:3000/_/api/1.0/signup.json: dial tcp [::1]:3000: getsockopt: connection refused`

maxtaco commented:
@jacobhaven this looks great.  We don't have a full story for third parties running our integration tests. We haven't made the server API code public yet, so you can't run a copy on localhost, as we do on our dev machines. However, we still have a staging server that we should open up for people to run tests one-off.

Go leads us to be a little bit stingy in exporting data structures, since the linter will complain if we don't comment them, and too may exported structures gum up the godocs. But your test makes a good case that we should export the packet specifications.

One last design issue: the word `anonymous` might be a little overloaded here, since there's a different capability in saltpack for the sender to hide his identity even to the receivers (also unexposed as of your).  It's an orthogonal feature to the one you implemented, so I wonder if the flag should be called `--hide-recipients` rather than `--anonymous`.

maxtaco commented on go/client/cmd_encrypt.go:
```diff
@@ -21,6 +21,7 @@ type CmdEncrypt struct {
 	recipients    []string
 	noSelfEncrypt bool
 	binary        bool
+	anonymous     bool
```
rename to `hideRecipients`?

maxtaco commented on go/client/cmd_encrypt.go:
```diff
@@ -55,6 +56,10 @@ func NewCmdEncrypt(cl *libcmdline.CommandLine, g *libkb.GlobalContext) cli.Comma
 				Usage: "Specify an outfile (stdout by default).",
 			},
 			cli.BoolFlag{
+				Name:  "a, anonymous",
```
Rename as above?

oconnor663 replied:
I'd also prefer to avoid picking a short flag for now, if that's ok with you guys. I prefer not to use short flags that have a common meaning, unless our meaning matches the convention. (`-a` would be "all" or "armor". `-r` would be "recursive" or (standard) "recipient". `-h` would be "help" or maybe "human-readable".)

jacobhaven replied:
Seems reasonable.

maxtaco commented on go/engine/saltpack_encrypt_test.go:
```diff
@@ -49,6 +51,72 @@ func TestSaltpackEncrypt(t *testing.T) {
+	PayloadKeyBox []byte `codec:"payloadkey"`
+}
+
+type encryptionHeader struct {
```
I think it makes sense to export these fields from `saltpack` if you need them here.

maxtaco commented on go/libkb/saltpack_enc.go:
```diff
@@ -15,6 +15,7 @@ type SaltpackEncryptArg struct {
 	Receivers []NaclDHKeyPublic
 	Sender    NaclDHKeyPair
 	Binary    bool
+	Anonymous bool
```
Rename as above...

maxtaco commented on go/saltpack/packets.go:
```diff
@@ -28,7 +28,7 @@ type encryptionHeader struct {
 	Ephemeral       []byte         `codec:"ephemeral"`
 	SenderSecretbox []byte         `codec:"sendersecretbox"`
-	Receivers       []receiverKeys `codec:"rcvrs"`
+	Receivers       []receiverKeys `codec:"rcvrs,omitempty"`
```
This isn't a good idea for forwards compatibility, since we're using the `toarray` encoding format (rather than named fields).  We had an intense internal debate as to which we should use, but in the end decided that arrays were safer since we wouldn't have to deal with decodes being fooled by dictionaries that repeat keys.

oconnor663 replied:
I don't think this `Receiver` list will/should ever be empty. That would mean that a message actually has no recipients -- there would be no PayloadKeyBoxes to open. It's the `receiverKeys.ReceiverKID` fields _inside_ the list that get emptied out in the anonymous case. We spec'd that the empty `ReceiverKID` should be a MessagePack `nil` value, so we might need to do some more plumbing to make that happen; I'll take a look.

I think I still prefer `nil` to an empty array, because it makes it less likely that a client could ignore the length, blindly pass the short array to NaCl, and start reading uninitialized memory. I've actually seen [_Python_ code](https://github.com/saltstack/libnacl/issues/58) do that.

oconnor663 replied:
Ok, cool, when I inspect a message with anonymous recipients, I do indeed see `nil` in the `ReceiverKID` fields. @maxtaco is that just how our MessagePack codec treats empty slices?

maxtaco replied:
yes

jacobhaven replied:
This all sounds reasonable, I wasn't sure of your msgpack handling.

maxtaco commented:
So some small nits, otherwise looks great

oconnor663 commented on go/libkb/saltpack.go:
```diff
@@ -38,6 +38,29 @@ func (b naclBoxPublicKey) HideIdentity() bool {
+type hiddenNaclBoxPublicKey NaclDHKeyPublic
+
+func (b hiddenNaclBoxPublicKey) ToKID() []byte {
+	return b[:]
```
Do I understand right that this should never be called? Maybe we should make it `panic`?

jacobhaven replied:
I can't guarantee that here (maybe some other code changes in the future would break it). I would really prefer to "subclass" `naclBoxPublicKey`, but I couldn't find a clean way to do that in Go.

oconnor663 commented:
Looks good to me, with the nits above. Thanks for putting this together! So awesome.

jacobhaven commented:
@maxtaco @oconnor663 I think I've addressed all the comments and updated the integration test (Could you please run `go test -run "TestSaltpackEncryptHideRecipients" ./go/engine/`against your staging server?

maxtaco commented:
It almost worked, but a few things were wrong:

1. Jack's idea to panic in `ToKID` was ill-advised, since we need it for receiver de-duping in `checkReceivers`
1. You had to encode as binary, not armored.
1. You have to double-decode the header, since it's double-decoded.  We do this in the spec to allow for convenient computation of hashes over the header.

Otherwise, it worked!

Fixes are here: https://github.com/keybase/client/tree/jacobhaven-tweaks

jacobhaven commented:
@maxtaco *added those changes to the PR* Thanks :)

oconnor663 commented:
> Jack's idea to panic in ToKID was ill-advised

The _illest!_

jacobhaven commented:
How is your Travis configured such that `travis-ci/push` passes but `/travis-ci/pr` fails? That's quite odd...

maxtaco commented:
Seems like had a C/I bug, we're playing with scripts today.  I'll rerun that failed build.

cjb commented:
(Push tests your branch by itself, and PR tests the merge of your branch onto master, and it turns out we broke CI on master just now, so that's why PR fails but push doesn't. Fixed in master now, and the PR one is rerunning.)

jacobhaven commented:
@cjb @maxtaco I've rebased off of master, but there still appear to be issues with the CI. Docker isn't finding the `kbweb` image.

maxtaco commented:
We're scratching our heads on why Circle CI is in the mix now. We had that on about ~3 months ago but disabled it.

jacobhaven commented:
Circle CI is probably building because I'm pushing to my `jacobhaven/client` fork and not the main `keybase` organization one.

maxtaco commented:
@cjb do you have any thoughts on why this PR is wreaking havoc on CI?  I guess because it's being pushed from a different repo.  FWIW, the tests run fine on my local machine...

oconnor663 commented:
The [Travis run](https://travis-ci.org/keybase/client/jobs/109717294) includes the error `Error parsing reference: "/kbweb" is not a valid repository/tag`. It looks like that comes from the expression `$CI_HOST/kbweb` in our `.travis.yml`. I'm guessing we define `$CI_HOST` in our Travis settings, and no one else does?

oconnor663 commented:
I just pushed https://github.com/keybase/client/tree/jack/pull2023 with the same tip as this PR (8193d4cd101f4894c910873b39f681968e92634c). Hopefully this build will work: https://travis-ci.org/keybase/client/builds/109766916

oconnor663 commented:
Looks good from here?
